File: AIiOS/AIiOS/Services/ExemplarScorer.swift
import Foundation

/// Pure scoring math for the Needle-embedder router (docs/ROUTING.md §4).
/// No platform imports beyond Foundation so it compiles into the standalone
/// test target; mirrors Android ExemplarScorer.kt — the two must stay
/// semantically identical (any divergence is a routing bug, see the corpus
/// parity test).
///
/// Pipeline (matches the probe that produced the GO measurement):
///   1. normalize whitespace in every text before embedding,
///   2. center prompt and exemplar vectors by the config's fixed mean
///      (raw needle_embed cosines are anisotropic),
///   3. fact = max cosine vs the category's exemplars > threshold (strict).
enum ExemplarScorer {

    // Character.isWhitespace covers the Unicode White_Space set, matching
    // Kotlin's [\s\p{Z}] regex and, for realistic text, Python's str.split()
    // — the normalization used to bake the shipped config.
    static func normalize(_ text: String) -> String {
        text.split(whereSeparator: \.isWhitespace).joined(separator: " ")
    }

    static func center(_ v: [Float], mean: [Float]) -> [Float] {
        precondition(v.count == mean.count, "vector dim \(v.count) != mean dim \(mean.count)")
        return zip(v, mean).map { $0 - $1 }
    }

    static func cosine(_ a: [Float], _ b: [Float]) -> Double {
        var dot = 0.0, na = 0.0, nb = 0.0
        for i in 0..<a.count {
            let x = Double(a[i]), y = Double(b[i])
            dot += x * y
            na  += x * x
            nb  += y * y
        }
        guard na > 0, nb > 0 else { return 0.0 }   // zero-norm guard
        return dot / (na.squareRoot() * nb.squareRoot())
    }

    /// - Parameters:
    ///   - promptVec: raw embedding of the (normalized) prompt
    ///   - mean: centering mean from the exemplar config, nil = no centering
    ///   - exemplars: per-category exemplar vectors, ALREADY centered
    ///   - thresholds: per-category strict-greater max-cosine thresholds
    static func semanticFacts(
        promptVec: [Float],
        mean: [Float]?,
        exemplars: [String: [[Float]]],
        thresholds: [String: Double]
    ) -> [String: FactValue] {
        let pv = mean.map { center(promptVec, mean: $0) } ?? promptVec
        var facts: [String: FactValue] = [:]
        for (cat, thr) in thresholds {
            let best = (exemplars[cat] ?? []).map { cosine(pv, $0) }.max() ?? -1.0
            facts[cat] = .bool(best > thr)
        }
        return facts
    }
}

File: AIAndroid/app/src/main/kotlin/ai/ecoinference/app/router/ExemplarScorer.kt
package ai.ecoinference.app.router

import kotlin.math.sqrt

/**
 * Pure scoring math for the Needle-embedder router (docs/ROUTING.md §4).
 * No platform imports so it runs on the plain JVM test runner; mirrors iOS
 * ExemplarScorer.swift — the two must stay semantically identical (any
 * divergence is a routing bug, see the corpus parity test).
 *
 * Pipeline (matches the probe that produced the GO measurement):
 *   1. normalize whitespace in every text before embedding,
 *   2. center prompt and exemplar vectors by the config's fixed mean
 *      (raw needle_embed cosines are anisotropic),
 *   3. fact = max cosine vs the category's exemplars > threshold (strict).
 */
object ExemplarScorer {

    // [\s\p{Z}] matches the Unicode White_Space set Swift's
    // Character.isWhitespace uses (\p{Z} covers NBSP and friends, which
    // Java's Char.isWhitespace() excludes) and, for realistic text, Python's
    // str.split() — the normalization used to bake the shipped config.
    private val whitespaceRun = Regex("[\\s\\p{Z}]+")

    fun normalize(text: String): String =
        text.split(whitespaceRun).filter { it.isNotEmpty() }.joinToString(" ")

    fun center(v: FloatArray, mean: FloatArray): FloatArray {
        require(v.size == mean.size) { "vector dim ${v.size} != mean dim ${mean.size}" }
        return FloatArray(v.size) { v[it] - mean[it] }
    }

    fun cosine(a: FloatArray, b: FloatArray): Double {
        var dot = 0.0; var na = 0.0; var nb = 0.0
        for (i in a.indices) {
            dot += a[i].toDouble() * b[i]
            na  += a[i].toDouble() * a[i]
            nb  += b[i].toDouble() * b[i]
        }
        return if (na > 0 && nb > 0) dot / (sqrt(na) * sqrt(nb)) else 0.0
    }

    /**
     * @param promptVec  raw embedding of the (normalized) prompt
     * @param mean       centering mean from the exemplar config, null = no centering
     * @param exemplars  per-category exemplar vectors, ALREADY centered
     * @param thresholds per-category strict-greater thresholds
     */
    fun semanticFacts(
        promptVec: FloatArray,
        mean: FloatArray?,
        exemplars: Map<String, List<FloatArray>>,
        thresholds: Map<String, Double>,
    ): Map<String, FactValue> {
        val pv = if (mean != null) center(promptVec, mean) else promptVec
        return thresholds.mapValues { (cat, thr) ->
            val best = exemplars[cat].orEmpty()
                .maxOfOrNull { cosine(pv, it) } ?: -1.0
            FactValue.BoolValue(best > thr)
        }
    }
}
